packages/camera/camera_platform_interface/lib/src/platform_interface/camera_platform.dart

```diff
@@ -0,0 +1,115 @@
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:plugin_platform_interface/plugin_platform_interface.dart';
+import 'package:camera_platform_interface/src/method_channel/method_channel_camera.dart';
+
+import '../../camera_platform_interface.dart';
+
+/// The interface that implementations of camera must implement.
+///
+/// Platform implementations should extend this class rather than implement it as `camera`
+/// does not consider newly added methods to be breaking changes. Extending this class
+/// (using `extends`) ensures that the subclass will get the default implementation, while
+/// platform implementations that `implements` this interface will be broken by newly added
+/// [CameraPlatform] methods.
+abstract class CameraPlatform extends PlatformInterface {
+  /// Constructs a CameraPlatform.
+  CameraPlatform() : super(token: _token);
+
+  static final Object _token = Object();
+
+  static CameraPlatform _instance = MethodChannelCamera();
+
+  /// The default instance of [CameraPlatform] to use.
+  ///
+  /// Defaults to [MethodChannelCamera].
+  static CameraPlatform get instance => _instance;
+
+  /// Platform-specific plugins should set this with their own platform-specific
+  /// class that extends [CameraPlatform] when they register themselves.
+  static set instance(CameraPlatform instance) {
+    PlatformInterface.verifyToken(instance, _token);
+    _instance = instance;
+  }
+
+  /// Completes with a list of available cameras.
+  Future<List<CameraDescription>> availableCameras() {
+    throw UnimplementedError('availableCameras() is not implemented.');
+  }
+
+  /// Initializes the camera on the device and returns its textureId.
+  Future<int> initializeCamera(
+    CameraDescription cameraDescription, {
+    ResolutionPreset resolutionPreset,
+    bool enableAudio,
+  }) {
+    throw UnimplementedError('initializeCamera() is not implemented.');
+  }
+
+  /// Returns a Stream of [CameraEvent]s.
+  Stream<CameraEvent> cameraEventsFor(int textureId) {
+    throw UnimplementedError('videoEventsFor() has not been implemented.');
+  }
+
+  /// Captures an image and saves it to [path].
+  Future<void> takePicture(int textureId, String path) {
+    throw UnimplementedError('takePicture() is not implemented.');
+  }
+
+  /// Prepare the capture session for video recording.
+  Future<void> prepareForVideoRecording() {
+    throw UnimplementedError('prepareForVideoRecording() is not implemented.');
+  }
+
+  /// Start a video recording and save the file to [path].
+  ///
+  /// A path can for example be obtained using
+  /// [path_provider](https://pub.dartlang.org/packages/path_provider).
+  ///
+  /// The file is written on the flight as the video is being recorded.
+  /// If a file already exists at the provided path an error will be thrown.
+  /// The file can be read as soon as [stopVideoRecording] returns.
+  Future<void> startVideoRecording(int textureId, String path) {
```

Comment:
@ditman, unlike the `takePicture` method (which keeps the data in memory and only then writes it to a file on disk) the `startVideoRecording` method writes the video to a file directly while it is being recorded.

What would be the best way to handle this keeping in mind that we probably cannot directly write video to disk on the web? My current we of thinking is that the method_channel implementation will assume the native platform will write the video to a temporary file and when `stopVideoRecording` is called it will load the file from disk and converts it to an XFile instance (based on the file path).

For the web implementation we need to have some other concept, probably a in memory byte array? I am just wondering if this will work since the video files can grow quite big very fast. Also this would mean that a web implementation based on method channels will never work (since we cannot write to file).

How would you feel about this?

Reply:
Ah, gotcha! No need to pass a pre-initialized XFile, the native implementation can record to disk and return a path (from the method channel call), and then that path gets wrapped in an XFile (similarly to what the image_picker does now).

In web, we can keep the array in memory as they're being recorded, and when the recording is done, we can return an XFile with the bytes that were captured so far.